### PR TITLE
Added constructor return type for wasm32 target

### LINKED ISSUE #1822: Bindgen generates incorrect return types for C++ class constructors in wasm32-unknown-emscripten
Bindgen generates incorrect return types for C++ class constructors in `wasm32-unknown-emscripten`, which causes linker warnings or even errors (when linking with `-Wl,--fatal-warnings`)

### Input C/C++ Header
```C++
class Foo {
public:
  Foo(int x);
};

Foo::Foo(int x) { }
```

### custom.rs
```Rust
mod bindings;

pub extern "C" fn create_foo() {
    let _ = unsafe { bindings::Foo::new(0) }; 
}
```

### bindings.rs
```Rust
/* automatically generated by rust-bindgen */

#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct Foo {
    pub _address: u8,
}
#[test]
fn bindgen_test_layout_Foo() {
    assert_eq!(
        ::std::mem::size_of::<Foo>(),
        1usize,
        concat!("Size of: ", stringify!(Foo))
    );
    assert_eq!(
        ::std::mem::align_of::<Foo>(),
        1usize,
        concat!("Alignment of ", stringify!(Foo))
    );
}
extern "C" {
    #[link_name = "\u{1}_ZN3FooC1Ei"]
    pub fn Foo_Foo(this: *mut Foo, x: ::std::os::raw::c_int);
}
impl Foo {
    #[inline(never)] // I added inline(never) to make this reproducable
    pub unsafe fn new(x: ::std::os::raw::c_int) -> Self {
        let mut __bindgen_tmp = ::std::mem::MaybeUninit::uninit();
        Foo_Foo(__bindgen_tmp.as_mut_ptr(), x);
        __bindgen_tmp.assume_init()
    }
}

```

### Bindgen Invocation
```
$ em++ -c class.cpp -o class.o
$ bindgen class.cpp --generate constructors,types --output bindings.rs
$ rustc --emit=obj --crate-type=lib --target wasm32-unknown-emscripten custom.rs -o bindings.o
$ emcc bindings.o class.o
```

### Actual Results
```
wasm-ld: warning: function signature mismatch: _ZN3FooC1Ei
>>> defined as (i32, i32) -> void in bindings.o
>>> defined as (i32, i32) -> i32 in class.o
```

### Expected Results
No warnings/errors. 

It appears that C++ constructors have an internal return value (which seemingly equals to `this`). However bindgen assumes that C++ constructors always return `void`, which causes function signature mismatches when linking code generated with bindgen with actual C++ object files.

I encountered this issue when trying to port `skia-safe` (which uses bindgen) to `wasm32-unknown-emscripten`.

### PR DESCRIPTION
**Pull request by tuxmark5**
PR to fix #1822

### CODE REVIEW

File: src/ir/function.rs [hunk 1]
```diff
@@ -495,7 +495,15 @@ impl FunctionSig {
         } else {
             ty.ret_type().ok_or(ParseError::Continue)?
         };
-        let ret = Item::from_ty_or_ref(ty_ret_type, cursor, None, ctx);
+
+        let ret = if ctx.is_target_wasm32() && is_constructor {
```
Comment (emilio) [inline src/ir/function.rs:499]: Can we switch the order here, `is_constructor` is cheaper to check, and should be false most of the time.

Also, can we add a test for this? Should be doable passing `--target` in the `bindgen-args` like `tests/headers/win32-thiscall_1_0.hpp` and others do.

File: src/ir/context.rs [hunk 2]
```diff
@@ -577,6 +577,14 @@ If you encounter an error missing from this list, please file an issue or a PR!"
         }
     }
 
+    /// Returns `true` if the target architecture is wasm32
+    pub fn is_target_wasm32(&self) -> bool {
```
Comment (emilio) [inline src/ir/context.rs:581]: This won't work if building with clang 4 or 3.9 (target_info will be null), but maybe that's fine.

**Review by emilio — CHANGES_REQUESTED** [review]
Thanks! code-wise looks good, but we need to have a test to avoid regressing this.

It should be easy, but if it's not for some reason let me know and I can probably merge with the tweak suggested below.

File: src/ir/context.rs (hunk 2 above)
Comment (tuxmark5) [reply src/ir/context.rs:581]: I doubt that wasm backend even existed for those versions of clang and as such I don't think what would be an issue.

File: src/ir/function.rs (hunk 1 above)
Comment (tuxmark5) [reply src/ir/function.rs]: Added a commit which addresses both of these comments.

File: src/ir/context.rs (hunk 2 above)
Comment (emilio) [reply src/ir/context.rs:581]: We still need to update the test expectations for these two versions, as we run CI on those. To do that you need to add the file in `tests/expectations/tests/libclang-3.9` (and same for 4).

With that bit, this patch looks great, thank you!

**Review by emilio — APPROVED** [review]
Looks great, thank you!

**Pull request merged**